### PR TITLE
Add Early Cycle EnKF workflow

### LINKED ISSUE #1021: Add Early Cycle EnKF
**Expected behavior**

workflow can generate xml file when EUPD_CYC="BOTH" or "gfs" when setup the experiment using "cycled" mode

**Current behavior**

Failed to generate xml file

**Machines affected**

Hera

**To Reproduce**

set EUPD_CYC="BOTH" and setup experiment using "cycled" mode

**Detailed Description**

When users setup a experiment using "cycled" mode and set EUPD_CYC="BOTH"  or "gfs", users can not create xml file. 

**Additional Information**

### PR DESCRIPTION
**Pull request by XianwuXue-NOAA**
**Description**
Updates workflow to allow early cycle EnKF to run.

Refs: #1021

**Type of change**
- [x] Bug fix (non-breaking change which fixes an issue)

**How Has This Been Tested?**
- [x] Cycled test on Hera
  
**Checklist**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code

### CODE REVIEW

File: workflow/applications.py
Comment (WalterKolczynski-NOAA) [inline workflow/applications.py:412]: Is this really the correct solution? If these jobs shouldn't run for gfs, then I think they should be moved to the hybrid_gdas_tasks list instead of hybrid_gdas_or_gfs_tasks. But I'm not 100% sure we don't need these jobs for early cycle. @aerorahul?

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

File: workflow/rocoto/workflow_tasks.py
Comment (WalterKolczynski-NOAA) [inline workflow/rocoto/workflow_tasks.py]: Just delete these instead of commenting them out.

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

File: workflow/rocoto/workflow_tasks.py
Comment (XianwuXue-NOAA) [reply workflow/rocoto/workflow_tasks.py]: Thanks. Yes, I will delete them later.

**Comment by WalterKolczynski-NOAA** [comment]
@aerorahul Other than the commented out lines that need to be removed, is there anything holding this up?

**Comment by XianwuXue-NOAA** [comment]
@WalterKolczynski-NOAA  Still working on it. We may change it back (see the last comment), when everything is done, I will delete all commented lines.

**Comment by XianwuXue-NOAA** [comment]
@aerorahul I have submitted my changes to this PR

File: ush/calcanl_gfs.py
Comment (XianwuXue-NOAA) [inline ush/calcanl_gfs.py]: I am not sure whether this change will affect others, so I keep the organize statement. After we make the final decision, I will clean this up.

File: jobs/JGDAS_ENKF_ECEN
Comment (WalterKolczynski-NOAA) [inline jobs/JGDAS_ENKF_ECEN:83]: Will this not mess up late-cycle?

File: jobs/JGDAS_ENKF_ECEN
Comment (WalterKolczynski-NOAA) [inline jobs/JGDAS_ENKF_ECEN:99]: Same potential issue here.

File: jobs/JGDAS_ENKF_SFC
Comment (WalterKolczynski-NOAA) [inline jobs/JGDAS_ENKF_SFC:100]: And here.

File: ush/calcanl_gfs.py
Comment (WalterKolczynski-NOAA) [reply ush/calcanl_gfs.py]: I think this may break IAU

File: workflow/rocoto/workflow_tasks.py
Comment (WalterKolczynski-NOAA) [inline workflow/rocoto/workflow_tasks.py]: Delete dead code.

File: workflow/rocoto/workflow_tasks.py
Comment (WalterKolczynski-NOAA) [inline workflow/rocoto/workflow_tasks.py]: More dead code. Not going to list them all.

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]
I'm not sure these changes leave late cycle functionality intact.

File: jobs/JGDAS_ENKF_ECEN
Comment (XianwuXue-NOAA) [reply jobs/JGDAS_ENKF_ECEN:83]: I am not sure, however, I think we can add some if statements to make it better for both early and late enkf

File: workflow/rocoto/workflow_tasks.py
Comment (XianwuXue-NOAA) [reply workflow/rocoto/workflow_tasks.py]: Just keep is as reference, and will delete dead code later.

File: workflow/rocoto/workflow_tasks.py
Comment (XianwuXue-NOAA) [reply workflow/rocoto/workflow_tasks.py]: Yes, just keeping them as ref and will delete them all later

**Comment by XianwuXue-NOAA** [comment]
@CatherineThomas-NOAA

- expdir on Hera: /scratch2/NCEPDEV/ensemble/noscrub/Xianwu.Xue/GEFS_v13_da/test_early_enkf_cold_BOTH_c96_test # my experiment is using C96/C48 with 20 members.
- Output: /scratch1/NCEPDEV/stmp2/Xianwu.Xue/o/test_early_enkf_cold_BOTH_c96_test
- ICs: /scratch2/NCEPDEV/ensemble/noscrub/Xianwu.Xue/GEFS_v13_da/z_WalterData/C96_cold_IC[C384_cold_IC]

File: jobs/JGDAS_ENKF_ECEN
Comment (TravisElless-NOAA) [reply jobs/JGDAS_ENKF_ECEN:83]: This change still allows both early and late cycle to point to previous gdas directory correctly

File: jobs/JGDAS_ENKF_SFC
Comment (TravisElless-NOAA) [inline jobs/JGDAS_ENKF_SFC]: Make sure $CDUMP here in line 97 and in line 83 are also updated to $GDUMP to point the early-cycle to the correct directory

File: jobs/JGDAS_ENKF_ECEN
Comment (XianwuXue-NOAA) [reply jobs/JGDAS_ENKF_ECEN:83]: Done. Thanks @WalterKolczynski-NOAA @TravisElless-NOAA

**Comment by XianwuXue-NOAA** [comment]
@WalterKolczynski-NOAA @TravisElless-NOAA I have completed one and half tests successfully. The log files are in : /scratch1/NCEPDEV/stmp2/Xianwu.Xue/o/test_early_enkf_cold_BOTH_c96/logs/2022012112. I am working on solving the issues which caused failure in lint.

**Comment by XianwuXue-NOAA** [comment]
I have fix the lint error.

**Comment by XianwuXue-NOAA** [comment]
@WalterKolczynski-NOAA @TravisElless-NOAA @CatherineThomas-NOAA @aerorahul Do you have any comments about this PR?  If you have, please let me know. I will merge the latest develop branch and then do the final test.

File: ush/calcanl_gfs.py
Comment (WalterKolczynski-NOAA) [reply ush/calcanl_gfs.py]: I guess this isn't needed, so it should be deleted.

**Comment by WalterKolczynski-NOAA** [comment]
@XianwuXue-NOAA If the DA group has signed off on the results, the only thing I'm waiting on is removal of the dead code.

**Comment by XianwuXue-NOAA** [comment]
@WalterKolczynski-NOAA Thanks. I will remove them today after the meeting this afternoon.

File: ush/calcanl_gfs.py
Comment (XianwuXue-NOAA) [reply ush/calcanl_gfs.py]: For this one, I think I can remove the "if" since else has the same statement with "if"

**Review by WalterKolczynski-NOAA — APPROVED** [review]

**Comment by XianwuXue-NOAA** [comment]
@WalterKolczynski-NOAA I have merged the latest develop branch into my branch and also delete the dead codes. Travis will triple check the changes and results.

**Comment by WalterKolczynski-NOAA** [comment]
Waiting on @CatherineThomas-NOAA to give final approval before merging.

**Review by TravisElless-NOAA — APPROVED** [review]
Tests show the late cycle reproduces the develop branch results and the early cycle points to the appropriate directories. From my perspective these changes should be good to go.

**Review by CatherineThomas-NOAA — APPROVED** [review]
Code looks good to me.  Thanks @XianwuXue-NOAA  and @TravisElless-NOAA for testing.  Approved.

**Comment by XianwuXue-NOAA** [comment]
@TravisElless-NOAA @CatherineThomas-NOAA and @WalterKolczynski-NOAA Thanks a lot for your approval.

**Pull request merged**
